### ISSUE TITLE
[Java] The offset buffer of empty BaseVariableWidthVector should not be empty when exposed through C Data Interface

### ISSUE THREAD

**Issue by viirya** [apache/arrow#40038]
### Describe the bug, including details regarding any error messages, version, and platform.

We encountered an error when exchanging string array from Java to Rust through Arrow C data interface. At Rust side, it complains that the buffer at position 1 (offset buffer) is null. After tracing down and some debugging, it looks like the issue is Java Arrow `BaseVariableWidthVector` class assigns an empty offset buffer if the array is empty (value count 0).

According to Arrow [spec](https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout) for variable size binary layout:

> The offsets buffer contains length + 1 signed integers ...

So for an empty string array, its offset buffer should be a buffer with one element (generally it is `0`).

### Component(s)

Java

**Comment by vibhatha** [comment]
@viirya Thanks for reporting this issue. Planning to make a PR?

**Comment by viirya** [comment]
@vibhatha Yea, I'm working on a fix locally. But it causes a few tests failed now. Still looking into fixing the tests.

**Comment by vibhatha** [comment]
Wonderful!

**Comment by viirya** [comment]
It looks like both cases are acceptable (empty or one single zero value element):

https://lists.apache.org/thread/w7g1zfqrjxx0bvrct0mt5zwxvdnc9nob

Close this for now.

**Comment by viirya** [comment]
Hmm, the code producing the issue is complicated. We execute TPCDS query in Spark/DataFusion and pass the results as Arrow vectors through C Data interface to Rust (arrow-rs).

I saw there are some tests in `c` module `RoundtripTest` . I'm looking to have a reproducer test there. But I cannot build `c` module on Mac...

**Comment by viirya** [comment]
Ah, after deeper debugging today, I found we were misled by `sun.misc.Unsafe.allocateMemory` API document.

> Allocates a new block of native memory, of the given size in bytes. ... The resulting native pointer will never be zero,...

But as I debugged it today, it actually returns zero value pointer. So it causes the Rust arrow think of it a NULL pointer.

And, there is an update to the API document:
https://github.com/openjdk/jdk/commit/209d87a856b1a7bd60910b517d8ff5beb322ec0b

It is changed to:

> The resulting native pointer will be zero if and only if the requested size is zero.

So once `UnsafeAllocationManager` is used , these empty buffers could be treated as NULL pointers actually when they are exported.

cc @lidavidm @pitrou

**Comment by viirya** [comment]
Actually we have `RoundtripTest` in `c` module which tests empty `ListVector`. But we don't have `c` unit tests with unsafe memory allocation manager like `vector` module. So it is not caught.

**Comment by viirya** [comment]
I am going to add unsafe run for `c` module in the PR #40043 too.

**Comment by lidavidm** [comment]
You could statically allocate a buffer in the allocator itself to represent the zero size buffer (in fact doesn't the allocator already do this?)

**Comment by viirya** [comment]
Zero size buffer? Why we need zero size buffer? Do you mean one zero value buffer?

**Comment by lidavidm** [comment]
Er, weren't we talking about this?

> But as I debugged it today, it actually returns zero value pointer. So it causes the Rust arrow think of it a NULL pointer.

**Comment by lidavidm** [comment]
I don't see why the allocator would give NULL for an allocation of size 4 or 8...

**Comment by viirya** [comment]
> Er, weren't we talking about this?

As you said, we already have it now. But for C Data Interface, as we discussed yesterday, it is not valid under C Data Interface. I think now we are going to send one value (zero) buffer instead of zero size buffer, isn't?

> I don't see why the allocator would give NULL for an allocation of size 4 or 8...

To clarify, we currently send a zero size buffer (i.e, `empty`) buffer through C Data Interface (not size 4 or 8)

For the reason, why the allocator give NULL for an allocation of empty buffer, see my previous comment: https://github.com/apache/arrow/issues/40038#issuecomment-1940350746.

In short, for the `empty` buffer allocated in `UnsafeAllocationManager`, which is a NULL pointer, because `MemoryUtil.UNSAFE.allocateMemory(0)` returns NULL pointer for 0 bytes allocation.

**Comment by viirya** [comment]
Let me try to summarize the issue:

1. Java Arrow uses empty buffer internally for offset buffer (valid, for historical reason). For C Data Interface, it is not valid. We are going to send one value (zero) buffer at the moment.
2. Currently the empty buffer sent by Java Arrow would be NULL pointer if unsafe allocator is used. It is because UNSAFE `allocateMemory` API returns a NULL pointer for zero bytes allocation.

**Comment by lidavidm** [comment]
Ok. (1) sounds fine. (2) is also fine. That's explicitly allowed so long as the buffer size really is 0. It won't be relevant after (1) is fixed, right?

**Comment by viirya** [comment]
Yea, if we don't send empty buffer as offset buffer, it is fine.
